### ISSUE TITLE
机器学习K线形态识别 | Candlestick Pattern Recognition

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4412]
**标签**: quantitative-strategy, machine-learning, cnn
### 摘要
使用卷积神经网络识别K线形态，如锤子线、吞没形态、十字星等。
### 算法描述
**核心原理：**
将K线图表转换为图像格式，使用CNN网络识别经典K线形态。

**关键步骤：**
1. OHLC数据转换为固定尺寸图像
2. 使用移动窗口生成样本
3. CNN提取空间特征
4. 分类器输出形态概率

**网络结构：**
Conv2D(32) → MaxPool → Conv2D(64) → Flatten → Dense(128) → Softmax
### Python代码示例
```python
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

class CandlestickImageGenerator:
    def __init__(self, window_size=20, img_height=32, img_width=32):
        self.window_size = window_size
        self.img_height = img_height
        self.img_width = img_width
        
    def ohlc_to_image(self, window):
        img = np.zeros((self.img_height, self.img_width, 3))
        h, w = self.img_height, self.img_width
        
        opens = (window['open'] - window['low']) / (window['high'] - window['low'] + 1e-8)
        highs = (window['high'] - window['low']) / (window['high'] - window['low'] + 1e-8)
        lows = (window['low'] - window['low']) / (window['high'] - window['low'] + 1e-8)
        closes = (window['close'] - window['low']) / (window['high'] - window['low'] + 1e-8)
        
        for i, (_, row) in enumerate(window.iterrows()):
            col = int(i / self.window_size * w)
            if col >= w: col = w - 1
            o = int(opens.iloc[i] * (h-1))
            h_price = int(highs.iloc[i] * (h-1))
            l_price = int(lows.iloc[i] * (h-1))
            c = int(closes.iloc[i] * (h-1))
            
            for j in range(l_price, min(h_price+1, h)):
                img[j, col] = [1, 1, 1]
        
        return img
    
    def label_patterns(self, window):
        o, h, l, c = window['open'].iloc[-1], window['high'].max(), window['low'].min(), window['close'].iloc[-1]
        body = abs(c - o)
        upper_shadow = h - max(c, o)
        lower_shadow = min(c, o) - l
        
        if body < 0.1 * (h - l) and upper_shadow > 2 * body and lower_shadow > 2 * body:
            return 'hammer'
        elif body < 0.1 * (h - l) and lower_shadow > 2 * body and upper_shadow > 2 * body:
            return 'inverted_hammer'
        elif c > o and body > 0.7 * (h - l):
            return 'bullish_engulfing'
        elif c < o and body > 0.7 * (h - l):
            return 'bearish_engulfing'
        return 'doji'

class SimpleCNNClassifier:
    def __init__(self, num_classes):
        self.num_classes = num_classes
        self.weights = None
        
    def fit(self, X, y, epochs=10, lr=0.01):
        n_samples, n_features = X.shape
        self.weights = np.random.randn(n_features, num_classes) * 0.01
        bias = np.zeros((1, num_classes))
        
        for epoch in range(epochs):
            indices = np.random.permutation(n_samples)
            for idx in indices:
                xi = X[idx:idx+1]
                yi = y[idx]
                
                logits = xi @ self.weights
                probs = np.exp(logits) / np.exp(logits).sum()
                
                error = probs.copy()
                error[yi] -= 1
                
                self.weights -= lr * xi.T @ error
                
        return self
    
    def predict(self, X):
        logits = X @ self.weights
        return np.argmax(logits, axis=1)

if __name__ == '__main__':
    np.random.seed(42)
    n = 1000
    data = pd.DataFrame({
        'open': np.random.uniform(95, 105, n),
        'high': lambda x: x + np.random.uniform(0, 10, n),
        'low': lambda x: x - np.random.uniform(0, 10, n),
        'close': np.random.uniform(95, 105, n)
    })
    data['high'] = data[['open', 'close']].max(axis=1) + np.random.uniform(0, 5, n)
    data['low'] = data[['open', 'close']].min(axis=1) - np.random.uniform(0, 5, n)
    
    generator = CandlestickImageGenerator(window_size=20)
    labels = []
    images = []
    
    for i in range(len(data) - 20):
        window = data.iloc[i:i+20]
        img = generator.ohlc_to_image(window)
        label = generator.label_patterns(window)
        labels.append(label)
        images.append(img.flatten())
    
    X = np.array(images)
    le = LabelEncoder()
    y = le.fit_transform(labels)
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
    
    clf = SimpleCNNClassifier(num_classes=len(le.classes_))
    clf.fit(X_train, y_train, epochs=20)
    
    acc = (clf.predict(X_test) == y_test).mean()
    print(f"测试准确率: {acc:.2%}")
    print(f"类别: {le.classes_}")
```
### 数据来源
- Reddit: r/algorithmictrading
- 知乎: K线形态、CNN图像识别
